Added response samples for most API endpoints

ariessa opened the pull request:
## Description
<!-- Concisely describe what the pull request does. -->
+ Added .gitattributes file
+ Removed POST /link and GET /link/{path} from documentation
+ Added response samples for the following API endpoints:
   - POST /auth_tokens/register
   - GET /images
   - GET /images/{identifier}
   - POST /images/{identifier}/report
   - GET /oembed
   - GET /recommendations/images/{identifier}
   - GET /sources
   - GET /rate_limit

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the *default* branch of the repository (`main` or `master`).
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [X] I added tests for the changes I made (if applicable).
- [X] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

ariessa committed:
Added response samples

ariessa committed:
Added response sample for recommendations

ariessa committed:
Added response sample for image_stats

ariessa committed:
Added description for key_info

ariessa committed:
Removed whitespaces

ariessa committed:
Fixed indentation

ariessa committed:
Fixed line too long

ariessa committed:
Fixed line too long in response samples

ariessa committed:
Fixed line too long again

ariessa committed:
Fixed line too long

ariessa committed:
Fixed style

ariessa committed:
Fixed trailing whitespace

aldenstpage commented on cccatalog-api/cccatalog/urls.py:
```diff
@@ -124,8 +124,8 @@
     ),
-    path('link', CreateShortenedLink.as_view(), name='make-link'),
-    path('link/<str:path>', ResolveShortenedLink.as_view(), name='resolve'),
+    # path('link', CreateShortenedLink.as_view(), name='make-link'),
```
The reason that the tests are failing is because these two endpoints have been commented out – rather than disable the endpoint entirely, for now, what we ought to do is hide the documentation for them to discourage further use. You can achieve this by setting the `swagger_schema = False` attribute on the right classes in `link_views.py` (see `site_views.Healthcheck` for an example).

I would like to shut down the link shortener views entirely but we have to make sure they're not in use anywhere first.

ariessa replied:
Understood and fixed

aldenstpage commented on cccatalog-api/cccatalog/api/views/image_views.py:
```diff
@@ -58,28 +59,89 @@ def _get_user_ip(request):
+                            "title": "File:Well test separator.svg",
+                            "id": "36537842-b067-4ca0-ad67-e00ff2e06b2d",
+                            "creator": "en:User:Oil&GasIndustry",
+                            "creator_url": "https://en.wikipedia.org/wiki/User:Oil%26GasIndustry",  # noqa
```
As you've discovered, there's a style test requirement that lines don't exceed a certain length. This helps us keep our Python code simpler, but it gets in the way when we're trying to show big JSON responses. Instead of using `# noqa`  silence each individual line that exceeds the limit, it would be better to use a broader exemption for the entire example response.

I think you can achieve this by adding `# noqa` after the closing brace for the image_search_response definition.

```
image_search_response = {
    . . .
} # noqa: E501
```

(`E501` is the [name of the line length violation](https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes) in the PEP 8 spec; it's best practice to only break one rule at a time)

ariessa replied:
I added `# noqa: E501` after the closing brace for the image_search_response definition but I got errors in the style test.

![image](https://user-images.githubusercontent.com/10581923/95604660-7df67300-0a8a-11eb-88c0-0517494871ec.png)

aldenstpage replied:
That's unfortunate, it seems like pycodestyle only allows you to ignore individual strings and lines. I did some research and it looks like the only option is to exclude the file from pycodestyle entirely, which we wouldn't really want to do to something as central as the view code.

Given those constraints, here's what I recommend:

1. Create a separate file for your JSON response examples in `cccatalog-api/cccatalog/example_responses.py`
2. Configure `pycodestyle` to ignore this file. See this line:       run: pycodestyle cccatalog-api/cccatalog --exclude='cccatalog-api/cccatalog/api/migrations' --max-line-length=80 --ignore=E402,E702
3. Import examples in each view.

That way, we can ignore style violations in the JSON examples specifically without impacting other parts of the code.

ariessa replied:
If I were to follow your recommendation, I'd have to import class `AboutImageResponse` into `cccatalog-api/cccatalog/example_responses.py`. Isn't cyclic dependency a bad practice?

https://github.com/creativecommons/cccatalog-api/blob/e976f93207e0d4fabd24dd546fd06601484cda1c/cccatalog-api/cccatalog/api/views/site_views.py#L50-L87

Also, I figured out that I can just separate the value into multiple lines.

https://github.com/creativecommons/cccatalog-api/blob/e976f93207e0d4fabd24dd546fd06601484cda1c/cccatalog-api/cccatalog/api/views/site_views.py#L171-L186

aldenstpage replied:
Yes, cyclic imports will typically mess things up in Python and cause a number of headaches. However, I think you can avoid importing anything into `example_responses.py` if you exclusively store the contents of the `openapi.Response(examples={...})` parameter and nothing else.

Line breaking the dict values does follow the letter of pep8 but not the spirit; I find it harder to read the examples inline after making that change

ariessa replied:
Alright, updated.

aldenstpage reviewed (dismissed):
Just a few minor changes to make here, looks like a great start overall!

ariessa committed:
Fixed PR

- Disabled link endpoints
- Replaced # noqa with # noqa: E501
- Added response sample for GET /rate_limit
- Added .gitattributes file

ariessa committed:
Added gitattributes file

ariessa committed:
SyntaxError in class ReportImageView

ariessa committed:
Added response sample for images_report_create

ariessa committed:
Fixed code style

ariessa committed:
Fix style again

ariessa committed:
Updates response sample

aldenstpage reviewed (approved):
@ariessa, this is really excellent, thanks for taking the time to work through the suggested changes.
